src/algebra/group/basic.lean

```diff
@@ -651,3 +651,27 @@ lemma bit1_sub [has_one M] (a b : M) : bit1 (a - b) = bit1 a - bit0 b :=
 (congr_arg (+ (1 : M)) $ bit0_sub a b : _).trans $ sub_add_eq_add_sub _ _ _
 
 end subtraction_comm_monoid
+
+/-- If a binary function from a type equipped with a total relation `r` to a monoid is
+  anti-symmetric (i.e. satisfies `f a b * f b a = 1`), in order to show it is multiplicative
+  (i.e. satisfies `f a c = f a b * f b c`), we may assume `r a b` and `r b c` are satisfied. -/
+@[to_additive additive_of_is_total "If a binary function from a type equipped with a total relation
+  `r` to an additive monoid is anti-symmetric (i.e. satisfies `f a b + f b a = 0`), in order to show
+  it is additive (i.e. satisfies `f a c = f a b + f b c`), we may assume `r a b` and `r b c`
+  are satisfied."]
+lemma multiplicative_of_is_total [monoid β] (f : α → α → β) (r : α → α → Prop) [t : is_total α r]
+  (hswap : ∀ a b, f a b * f b a = 1)
+  (hmul : ∀ {a b c}, r a b → r b c → f a c = f a b * f b c)
+  (a b c : α) : f a c = f a b * f b c :=
+begin
+  suffices : ∀ b c, r b c → f a c = f a b * f b c,
+  { obtain hbc | hcb := t.total b c,
+    { exact this b c hbc },
+    { rw [this c b hcb, mul_assoc, hswap c b, mul_one] } },
+  intros b c hbc,
+  obtain hab | hba := t.total a b,
+  { exact hmul hab hbc },
+  obtain hac | hca := t.total a c,
+  { rw [hmul hba hac, ← mul_assoc, hswap a b, one_mul] },
+  { rw [← one_mul (f a c), ← hswap a b, hmul hbc hca, mul_assoc, mul_assoc, hswap c a, mul_one] },
+end
```

Comment:
E.g., this version can be used for integrals:
```suggestion
@[to_additive additive_of_is_total']
lemma multiplicative_of_is_total' [monoid β] (f : α → α → β) (p r : α → α → Prop) [is_total α r]
  (hp : symmetric p) (hf_swap : ∀ a b, f a b * f b a = 1)
  (hmul : ∀ {a b c}, p a b → r a b → p b c → r b c → p a c → f a c = f a b * f b c)
  (a b c : α) (hab : p a b) (hbc : p b c) (hac : p a c) : f a c = f a b * f b c :=
begin
  suffices : ∀ {b c}, p a b → p b c → r b c → p a c → f a c = f a b * f b c,
  { obtain hbc' | hcb' := total_of r b c,
    { exact this hab hbc hbc' hac },
    { rw [this hac (hp hbc) hcb' hab, mul_assoc, hf_swap c b, mul_one] } },
  intros b c hab hbc hbc' hac,
  obtain hab' | hba' := total_of r a b,
  { exact hmul hab hab' hbc hbc' hac },
  rw [← one_mul (f a c), ← hf_swap a b, mul_assoc],
  obtain hac' | hca' := total_of r a c,
  { rw [hmul (hp hab) hba' hac hac' hbc] },
  { rw [hmul hbc hbc' (hp hac) hca' (hp hab), mul_assoc, hf_swap, mul_one] },
end

/-- If a binary function from a type equipped with a total relation `r` to a monoid is
  anti-symmetric (i.e. satisfies `f a b * f b a = 1`), in order to show it is multiplicative
  (i.e. satisfies `f a c = f a b * f b c`), we may assume `r a b` and `r b c` are satisfied. -/
@[to_additive additive_of_is_total "If a binary function from a type equipped with a total relation
  `r` to an additive monoid is anti-symmetric (i.e. satisfies `f a b + f b a = 0`), in order to show
  it is additive (i.e. satisfies `f a c = f a b + f b c`), we may assume `r a b` and `r b c`
  are satisfied."]
lemma multiplicative_of_is_total [monoid β] (f : α → α → β) (r : α → α → Prop) [t : is_total α r]
  (hswap : ∀ a b, f a b * f b a = 1)
  (hmul : ∀ {a b c}, r a b → r b c → f a c = f a b * f b c)
  (a b c : α) : f a c = f a b * f b c :=
multiplicative_of_is_total' f (λ _ _, true) r (λ _ _, id) hswap (λ _ _ _ _ h₁ _ h₂ _, hmul h₁ h₂) _
  _ _ trivial trivial trivial
```

Reply:
This saves only 1 line for the lemma about integrals:
```
lemma integral_add_adjacent_intervals (hab : interval_integrable f μ a b)
  (hbc : interval_integrable f μ b c) :
  ∫ x in a..b, f x ∂μ + ∫ x in b..c, f x ∂μ = ∫ x in a..c, f x ∂μ :=
begin
  convert (additive_of_is_total' (λ a b, ∫ x in a..b, f x ∂μ) (interval_integrable f μ) (≤)
    (λ a b h, h.symm) (λ a b, eq_neg_iff_add_eq_zero.1 $ integral_symm _ _)
    (λ a b c hab hab' hbc hbc' hac, _)
    a b c hab hbc (hab.trans hbc)).symm,
  simp only [integral_of_le hab', integral_of_le hbc', integral_of_le (hab'.trans hbc'),
    ← Ioc_union_Ioc_eq_Ioc hab' hbc'],
  exact integral_union Ioc_disjoint_Ioc_same measurable_set_Ioc hab.1 hbc.1
end

lemma integral_add_adjacent_intervals_cancel (hab : interval_integrable f μ a b)
  (hbc : interval_integrable f μ b c) :
  ∫ x in a..b, f x ∂μ + ∫ x in b..c, f x ∂μ + ∫ x in c..a, f x ∂μ = 0 :=
by rw [integral_add_adjacent_intervals hab hbc, integral_symm, neg_add_self]
```
but the proof no longer needs `set.Ioc_union_Ioc_union_Ioc_cycle`.

Reply:
Thanks! It's intended to be used in #18040 though. There we need to restrict to a subtype but has no condition like integrability; I'm considering adding a version requiring a,b,c\in s. Anyway the current version is already in mathlib4 so modifications to the statement would require a matching PR.

Reply:
I've incorporated the symmetric version and replace the original version by a predicate version deduced from the symmetric version. Ported in https://github.com/leanprover-community/mathlib4/pull/1519